### PR TITLE
pool: Exceptionally stop pool in case of a long negative yield period

### PR DESCRIPTION
**Pull request by axic**
This is mostly here for a discussion starter.

### CODE REVIEW

File: contracts/TempusPool.sol [hunk 1]
```diff
@@ -24,13 +24,16 @@ abstract contract TempusPool is ITempusPool {
...
+    uint256 public constant override maximumNegativeYieldDuration = 7 days;
+
+    uint256 public constant override version = 1;
 
     address public immutable override yieldBearingToken;
     address public immutable override backingToken;
 
     uint256 public immutable override startTime;
     uint256 public immutable override maturityTime;
+    uint256 public override exceptionalHaltTime = type(uint256).max;
```
Comment (axic) [inline contracts/TempusPool.sol:36]: This is a trick so that we don't need to check `!= 0` in `halted` below.

File: contracts/TempusPool.sol [hunk 2]
```diff
@@ -127,6 +133,10 @@ abstract contract TempusPool is ITempusPool {
         return block.timestamp >= maturityTime;
     }
 
+    function halted() public view override returns (bool) {
```
Comment (axic) [inline contracts/TempusPool.sol:136]: Not sure we need this helper -- we could incorporate this into `matured`.

File: contracts/TempusPool.sol [hunk 3]
```diff
@@ -454,6 +465,27 @@ abstract contract TempusPool is ITempusPool {
         return toBackingToken ? backingTokens : yieldTokens;
     }
 
+    /// @dev This updates the underlying pool's interest rate, and also updates the internal
+    ///      tracking of negative interest periods.
+    function updateAndCheckInterestRate() internal returns (uint256 rate) {
```
Comment (axic) [inline contracts/TempusPool.sol:470]: I wonder if this helper is the right approach, or doing some modifier/helper function and explicitly calling it on the entry points, or on `burnShares`/`mintShares`.

File: contracts/TempusPool.sol [hunk 4]
```diff
@@ -454,6 +465,27 @@ abstract contract TempusPool is ITempusPool {
...
+                lastNegativeYield = block.timestamp;
+            } else {
+                // Already in a negative yield period, exceeding the duration.
+                if ((lastNegativeYield + maximumNegativeYieldDuration) >= block.timestamp) {
+                    exceptionalHaltTime = block.timestamp;
+                }
+            }
+        } else {
+            // Reset period.
+            lastNegativeYield = 0;
```
Comment (axic) [inline contracts/TempusPool.sol:485]: This here in the generic case means resetting the storage value with 0.  This adds a constant 100 gas overhead.

Any better ideas?

File: contracts/TempusPool.sol [hunk 5]
```diff
@@ -47,15 +50,18 @@ abstract contract TempusPool is ITempusPool {
...
     uint256 public immutable override maxDepositFee;
     uint256 public immutable override maxEarlyRedeemFee;
     uint256 public immutable override maxMatureRedeemFee;
 
-    /// total amount of fees accumulated in pool
+    /// Total amount of fees accumulated in pool.
     uint256 public override totalFees;
 
+    /// Timestamp when the negative yield period was entered.
+    uint256 private lastNegativeYield;
```
Comment (axic) [inline contracts/TempusPool.sol:63]: Potentially could just rework this to merge `exceptionalHaltTime` and `lastNegativeYield` into a single variable.  In this case though we definitely need to check for `!= 0` in the matured/halted helpers.

File: contracts/TempusPool.sol (hunk 2 above)
Comment (axic) [reply contracts/TempusPool.sol]: Did that in a second commit.

File: contracts/TempusPool.sol [hunk 6]
```diff
@@ -24,13 +24,16 @@ abstract contract TempusPool is ITempusPool {
     using UntrustedERC20 for IERC20;
     using Fixed256xVar for uint256;
 
-    uint public constant override version = 1;
+    uint256 public constant override maximumNegativeYieldDuration = 7 days;
+
+    uint256 public constant override version = 1;
```
Comment (RedFox20) [inline contracts/TempusPool.sol:29]: Version should be the first field

File: contracts/TempusPool.sol [hunk 7]
```diff
@@ -454,6 +460,27 @@ abstract contract TempusPool is ITempusPool {
         return toBackingToken ? backingTokens : yieldTokens;
     }
 
+    /// @dev This updates the underlying pool's interest rate, and also updates the internal
+    ///      tracking of negative interest periods.
+    function updateAndCheckInterestRate() internal returns (uint256 rate) {
+        rate = updateInterestRate();
+
+        if (rate < initialInterestRate) {
+            if (lastNegativeYield == 0) {
```
Comment (RedFox20) [inline contracts/TempusPool.sol:469]: Names of the fields are a bit generic, it would be better to use specific names: `negativeYieldStartTime`

File: contracts/TempusPool.sol [hunk 8]
```diff
@@ -195,7 +201,7 @@ abstract contract TempusPool is ITempusPool {
         )
     {
         require(!matured(), "Maturity reached.");
```
Comment (RedFox20) [inline contracts/TempusPool.sol:203]: `updateAndCheckInterestRate()` will affect `matured()`, is this the intended ordering of calls? If so, it needs a comment now, otherwise, rate should be done before `matured()` check

File: contracts/TempusPool.sol [hunk 9]
```diff
@@ -289,7 +295,7 @@ abstract contract TempusPool is ITempusPool {
         PrincipalShare(address(principalShare)).burnFrom(from, principalAmount);
         YieldShare(address(yieldShare)).burnFrom(from, yieldAmount);
 
-        uint256 currentRate = updateInterestRate();
+        uint256 currentRate = updateAndCheckInterestRate();
```
Comment (RedFox20) [inline contracts/TempusPool.sol:298]: Same issue again, `updateAndCheckInterestRate()` will affect `matured()` state, so is the previous maturity check above correct or not?
Also, this change will automatically finalize the pool at some point due to negative yield

File: contracts/TempusPool.sol [hunk 10]
```diff
@@ -124,7 +130,7 @@ abstract contract TempusPool is ITempusPool {
         returns (uint256 backingTokenAmount);
 
     function matured() public view override returns (bool) {
-        return block.timestamp >= maturityTime;
+        return (block.timestamp >= maturityTime) || (block.timestamp >= exceptionalHaltTime);
```
Comment (yuval-weiss) [inline contracts/TempusPool.sol:133]: I think it would be better if we have 2 types of "paused" states - `matured` & `halted`. This makes more sense than merging these 2 conditions into the "matured" state imo.

File: contracts/TempusPool.sol (hunk 7 above)
Comment (axic) [reply contracts/TempusPool.sol]: Renamed.

File: contracts/TempusPool.sol (hunk 8 above)
Comment (axic) [reply contracts/TempusPool.sol:203]: I didn't wanted to change the order because didn't wanted to query the rate after maturity, and figured the extra check for negative yield afterwards would be sufficient -- so a comment is definitely worth it here.

However open for different designs, that `negative yield!` check feels like a bit of an afterthought.

File: test/TempusPool.Redeem.test.ts [hunk 11]
```diff
@@ -141,6 +141,33 @@ describeForEachPool("TempusPool Redeem", (pool:PoolTestFixture) =>
...
+  {
+    await pool.createDefault();
+    let [owner, user] = pool.signers;
+    await pool.setupAccounts(owner, [[user, 200]]);
+
+    await pool.depositYBT(user, 100, /*recipient:*/user);
+    (await pool.userState(user)).expect(100, 100, /*yieldBearing:*/100);
+
+    await pool.setTimeRelativeToPoolStart(24*60*60); // 1 day
+    await pool.setInterestRate(0.9);
```
Comment (axic) [inline test/TempusPool.Redeem.test.ts:154]: Oh, actually need to do an action here to trigger updates.

File: test/TempusPool.Redeem.test.ts [hunk 12]
```diff
@@ -141,6 +141,42 @@ describeForEachPool("TempusPool Redeem", (pool:PoolTestFixture) =>
...
+    await pool.setInterestRate(0.6);
+
+    await pool.redeemToYBT(user, 90, 90);
+    if (pool.yieldPeggedToAsset)
+    {
+        (await pool.userState(user)).expect(0, 0, /*yieldBearing:*/120);
+    }
+    else
+    {
+        (await pool.userState(user)).expect(0, 0, /*yieldBearing:*/200);
```
Comment (axic) [inline test/TempusPool.Redeem.test.ts:176]: This (and other values) needs to be reviewed, as the previous test case ("Should work before maturity with equal shares, with negative yield") arrives at this number too, despite this withdrawing prior.

What was initially confusing is that the accounts are funding with 200 and we only deposit 100.

File: test/TempusPool.Redeem.test.ts [hunk 13]
```diff
@@ -141,6 +141,42 @@ describeForEachPool("TempusPool Redeem", (pool:PoolTestFixture) =>
...
+  it.includeIntegration("Should work before maturity with equal shares, with long period of negative yield", async () =>
+  {
+    await pool.createDefault();
+    let [owner, user] = pool.signers;
+    await pool.setupAccounts(owner, [[user, 200]]);
+
+    await pool.depositYBT(user, 100, /*recipient:*/user);
+    (await pool.userState(user)).expect(100, 100, /*yieldBearing:*/100);
+
+    await pool.setTimeRelativeToPoolStart(24*60*60); // 1 day
```
Comment (yuval-weiss) [inline test/TempusPool.Redeem.test.ts:153]: This should be a relative number (range 0 to 1) to the start of the pool (0.5 to move time to 50% until maturity).
And also - `pool.createDefault` creates a pool with a maturity of 1 hour, so 1 day or 8 days is after maturity

File: contracts/TempusPool.sol (hunk 8 above)
Comment (yuval-weiss) [reply contracts/TempusPool.sol:203]: I definitely think the ordering here is problematic... It just means that the pool will never actually be halted on the same transaction where `exceptionalHaltTime` is set because there's no check for that after this variable is set. As I mentioned in another comment, I also think we should separate the `halted` state from the `matured` state (as those are currently merged together).

File: test/TempusPool.Redeem.test.ts (hunk 13 above)
Comment (axic) [reply test/TempusPool.Redeem.test.ts:153]: Ahh, good catch that it is a percentage. Should add a safety check into it.

File: contracts/TempusPool.sol [hunk 14]
```diff
@@ -280,6 +288,7 @@ abstract contract TempusPool is ITempusPool {
         require(IERC20(address(yieldShare)).balanceOf(from) >= yieldAmount, "Insufficient yields.");
 
         if (matured()) {
+            console.log("Matured", (exceptionalHaltTime != type(uint256).max) ? exceptionalHaltTime : 0);
```
Comment (yuval-weiss) [inline contracts/TempusPool.sol:291]: remove

File: contracts/TempusPool.sol (hunk 8 above)
Comment (axic) [reply contracts/TempusPool.sol]: Swapped in #439.

File: contracts/TempusPool.sol (hunk 9 above)
Comment (axic) [reply contracts/TempusPool.sol]: Swapped in #439.

File: contracts/TempusPool.sol [hunk 15]
```diff
@@ -450,6 +459,31 @@ abstract contract TempusPool is ITempusPool, ReentrancyGuard, Ownable {
         return toBackingToken ? backingTokens : yieldTokens;
     }
 
+    /// @dev This updates the underlying pool's interest rate, and also updates the internal
```
Comment (axic) [inline contracts/TempusPool.sol:462]: Based on #428 we may want to name this `updateAndValidateInterestRate` instead.

File: contracts/TempusPool.sol [hunk 16]
```diff
@@ -255,7 +263,7 @@ abstract contract TempusPool is ITempusPool, ReentrancyGuard, Ownable {
 
     function finalize() public override {
         if (matured() && maturityInterestRate == 0) {
-            maturityInterestRate = updateInterestRate();
+            maturityInterestRate = updateAndCheckInterestRate();
```
Comment (yuval-weiss) [inline contracts/TempusPool.sol:266]: I think this should just be `updateInterestRate` and `updateAndCheckInterestRate` should be used in `burnShares` (instead of `updateInterestRate`). If `finalize` is called externally it means the pool has matured already anyways...

File: contracts/TempusPool.sol (hunk 15 above)
Comment (yuval-weiss) [reply contracts/TempusPool.sol:462]: I think `updateAndValidateInterestRate` is better

**Comment by axic** [comment]
Need to check why Aave aDAI behaves differently:
```
  Aave aDAI <> TempusPool Redeem
    createFixture Aave|aDAI|1|1382400|0.1|0.02|5|5 1174ms
Resetting negative yield period
Entering negative yield period 1638611996
Resetting negative yield period
Entering negative yield period 1639130395
Check for long negative yield period
Exceptionally halting 1639650176
Matured 1639650176 0
Resetting negative yield period
Matured 1639650176 200000000000000000

    1) Should work before maturity with equal shares, with fluctuating yield                        2469ms

  Aave aUSDC <> TempusPool Redeem
    createFixture Aave|aUSDC|1|1382400|0.1|0.02|5|5 503ms
Resetting negative yield period
Entering negative yield period 1638776156
Resetting negative yield period
Entering negative yield period 1639787036
Check for long negative yield period
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         796ms

  Compound cDAI <> TempusPool Redeem
    createFixture Compound|cDAI|1|1382400|0.1|0.02|5|5 502ms
Resetting negative yield period
Entering negative yield period 1638932110
Resetting negative yield period
Entering negative yield period 1640410846
Check for long negative yield period
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         789ms

  Compound cUSDC <> TempusPool Redeem
    createFixture Compound|cUSDC|1|1382400|0.1|0.02|5|5 497ms
Resetting negative yield period
Entering negative yield period 1639080266
Resetting negative yield period
Entering negative yield period 1641003465
Check for long negative yield period
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         764ms

  Lido stETH <> TempusPool Redeem
    createFixture Lido|stETH|1|1382400|0.1|0.02|5|5 425ms
Resetting negative yield period
Entering negative yield period 1639221012
Check for long negative yield period
Check for long negative yield period
Check for long negative yield period
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         731ms

  Yearn yvDAI <> TempusPool Redeem
    createFixture Yearn|yvDAI|1|1382400|0.1|0.02|5|5 475ms
Resetting negative yield period
Entering negative yield period 1639354722
Resetting negative yield period
Entering negative yield period 1642101291
Check for long negative yield period
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         750ms

  Yearn yvUSDC <> TempusPool Redeem
    createFixture Yearn|yvUSDC|1|1382400|0.1|0.02|5|5 453ms
Resetting negative yield period
Entering negative yield period 1639481747
Resetting negative yield period
Entering negative yield period 1642609387
Check for long negative yield period
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         704ms
```

**Comment by axic** [comment]
After rebasing this on #445, which means there is no period exceeding 7 days of negative interest (before that by mistake the period was 7+ days), the results are the follow (lido fails, because it matures):
```
  Aave aDAI <> TempusPool Redeem
Initial interest rate 1000000000000000000
    createFixture Aave|aDAI|1|1382400|0.1|0.02|5|5 1129ms
Resetting negative yield period
Entering negative yield period 1638582680 900000000000000000
Resetting negative yield period
Entering negative yield period 1639187480 600000000000000000
Check for long negative yield period 1639792280 1639619480 200000000000000000
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         2232ms

  Aave aUSDC <> TempusPool Redeem
Initial interest rate 1000000000000000000
    createFixture Aave|aUSDC|1|1382400|0.1|0.02|5|5 500ms
Resetting negative yield period
Entering negative yield period 1638582680 900000000000000000
Resetting negative yield period
Entering negative yield period 1639187480 600000000000000000
Check for long negative yield period 1639792280 1639619480 200000000000000000
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         777ms

  Compound cDAI <> TempusPool Redeem
Initial interest rate 10000000000000000000000000000
    createFixture Compound|cDAI|1|1382400|0.1|0.02|5|5 491ms
Resetting negative yield period
Entering negative yield period 1638582682 9000000000000000000000000000
Resetting negative yield period
Entering negative yield period 1639187482 6000000000000000000000000000
Check for long negative yield period 1639792282 1639619482 2000000000000000000000000000
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         766ms

  Compound cUSDC <> TempusPool Redeem
Initial interest rate 10000000000000000
    createFixture Compound|cUSDC|1|1382400|0.1|0.02|5|5 485ms
Resetting negative yield period
Entering negative yield period 1638582683 9000000000000000
Resetting negative yield period
Entering negative yield period 1639187483 6000000000000000
Check for long negative yield period 1639792283 1639619483 2000000000000000
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         759ms

  Lido stETH <> TempusPool Redeem
Initial interest rate 1000000000000000000
    createFixture Lido|stETH|1|1382400|0.1|0.02|5|5 423ms
Resetting negative yield period
Entering negative yield period 1638582681 900000000000000000
Check for long negative yield period 1639187481 1638928281 999999999999999999
Check for long negative yield period 1639187481 1639187481 599999999999999999
Exceptionally halting 1639187481 599999999999999999
Matured 1639187481 0
Check for long negative yield period 1639187481 1639619481 199999999999999999
Exceptionally halting 1639619481 199999999999999999
Matured 1639619481 599999999999999999
Resetting negative yield period
Matured 1639619481 599999999999999999

    1) Should work before maturity with equal shares, with fluctuating yield                        718ms

  Yearn yvDAI <> TempusPool Redeem
Initial interest rate 1000000000000000000
    createFixture Yearn|yvDAI|1|1382400|0.1|0.02|5|5 481ms
Resetting negative yield period
Entering negative yield period 1638582683 900000000000000000
Resetting negative yield period
Entering negative yield period 1639187483 600000000000000000
Check for long negative yield period 1639792283 1639619483 200000000000000000
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         770ms

  Yearn yvUSDC <> TempusPool Redeem
Initial interest rate 1000000
    createFixture Yearn|yvUSDC|1|1382400|0.1|0.02|5|5 580ms
Resetting negative yield period
Entering negative yield period 1638582684 900000
Resetting negative yield period
Entering negative yield period 1639187484 600000
Check for long negative yield period 1639792284 1639619484 200000
Resetting negative yield period
    ✓ Should work before maturity with equal shares, with fluctuating yield                         891ms
```

**Comment by axic** [comment]
Ok I think the reason is that Lido's rate setting is not precise enough and instead of `1` it ends ups at `0.999999999999999999` which is still considered negative.

I changed the test cases to do `1.1` as the non-negative rate and now it seems to be correct.

File: contracts/TempusPool.sol [hunk 17]
```diff
@@ -447,6 +463,42 @@ abstract contract TempusPool is ITempusPool, ReentrancyGuard, Ownable {
         return toBackingToken ? backingTokens : yieldTokens;
     }
 
+    /// @dev This updates the underlying pool's interest rate, and also updates the internal
+    ///      tracking of negative interest periods.
+    function updateAndValidateInterestRate() internal returns (uint256 rate) {
+        rate = updateInterestRate();
+
+        // Short circuit. No need for the below after maturity.
+        if (matured()) {
```
Comment (axic) [inline contracts/TempusPool.sol:472]: This is a new feature, to avoid doing any calculation once matured.

File: contracts/TempusPool.sol [hunk 18]
```diff
@@ -28,11 +28,14 @@ abstract contract TempusPool is ITempusPool, ReentrancyGuard, Ownable {
 
     uint256 public constant override version = 1;
 
+    uint256 public constant override maximumNegativeYieldDuration = 7 days;
```
Comment (axic) [inline contracts/TempusPool.sol:31]: Should make this configurable, in a new PR.

File: contracts/ITempusPool.sol [hunk 19]
```diff
@@ -77,7 +77,16 @@ interface ITempusPool is ITempusFees {
     /// @return Maturity time of the pool.
     function maturityTime() external view returns (uint256);
 
+    /// @return Time of exceptional halting of the pool.
+    /// In case the pool is still in operation, this must return type(uint256).max.
+    function exceptionalHaltTime() external view returns (uint256);
+
+    /// @return The maximum allowed time (in seconds) to pass with negative yield.
+    function maximumNegativeYieldDuration() external view returns (uint256);
```
Comment (axic) [inline contracts/ITempusPool.sol:85]: Should add basic tests for these two.

File: test/utils/TempusPool.ts [hunk 20]
```diff
@@ -398,6 +398,27 @@ export class TempusPool extends ContractBase {
...
   }
 
+  /**
+   * @returns The exceptional halt time of the pool
+   * @note This returns null in case it is not set (i.e. has the special value of `type(uin256).max`)
+   */
+  async exceptionalHaltTime(): Promise<NumberOrString | null> {
+    let exceptionalHaltTime:BigNumber = await this.contract.exceptionalHaltTime();
+//    if (exceptionalHaltTime.eq(BigNumber.from(MAX_UINT256))) {
+    if (exceptionalHaltTime.toHexString() === MAX_UINT256) {
```
Comment (axic) [inline test/utils/TempusPool.ts:408]: Here are two options to accomplish this. Also could just use a local string constant instead of important this one.

File: contracts/TempusPool.sol (hunk 18 above)
Comment (axic) [reply contracts/TempusPool.sol:31]: See #449.

**Review by RedFox20 — DISMISSED** [review]

**Review by RedFox20 — APPROVED** [review]
